src/ring_theory/dedekind_domain/adic_valuation.lean

```diff
@@ -333,3 +333,40 @@ instance adic_completion.has_lift_t : has_lift_t K (v.adic_completion K) :=
 def adic_completion_integers : valuation_subring (v.adic_completion K) := valued.v.valuation_subring
 
 end is_dedekind_domain.height_one_spectrum
+
+variables (R K)
+
+lemma adic_completion.is_integer (x : v.adic_completion K) :
+  x ∈ v.adic_completion_integers K ↔ (valued.v x : ℤₘ₀) ≤ 1 :=
+iff.rfl
+
+/-- The natural inclusion of `R` in `adic_completion_integers`, as a ring homomorphism. -/
+@[simps] def inj_adic_completion_integers : R →+* v.adic_completion_integers K :=
+{ to_fun := λ r, ⟨coe $ algebra_map R K r,
+  begin
+    change @valued.extension K _ _ _ v.adic_valued (algebra_map R K r) ≤ 1,
+    rw @valued.extension_extends K _ _ _ v.adic_valued (algebra_map R K r),
+    exact v.valuation_le_one _,
+  end⟩,
```

Comment:
The reason for the `change` here is because there's no API for `valued.valued_completion`. I suggest we add:
```lean
@[simp, norm_cast] lemma valued_completion_apply (x : K) : valued.v (x : hat K) = v x :=
extension_extends x
```
just below [valued_completion](https://github.com/leanprover-community/mathlib/blob/28f7172e9677fd412cdd500c914d18d423846715/src/topology/algebra/valued_field.lean#L352) and then we can avoid using `change` here:
```suggestion
{ to_fun := λ r, ⟨coe $ algebra_map R K r, by simpa only [adic_completion.is_integer,
    valued.valued_completion_apply] using v.valuation_le_one _⟩,
```

Reply:
Thanks, I added this.